src/app/CommandSender.cpp

```diff
@@ -367,14 +398,46 @@ CHIP_ERROR CommandSender::ProcessInvokeResponseIB(InvokeResponseIB::Parser & aIn
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CommandSender::PrepareCommand(const CommandPathParams & aCommandPathParams, bool aStartDataStruct)
+CHIP_ERROR CommandSender::GetAdditionalInvokeResponseElements(AdditionalInvokeResponseElements & aInvokeReponseElements)
+{
+    VerifyOrReturnError(mState == State::ResponseReceived, CHIP_ERROR_INCORRECT_STATE);
+    aInvokeReponseElements = mAdditionalResponseElements;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CommandSender::SetBatchCommandsConfig(uint16_t aRemoteMaxPathsPerInvoke)
+{
+#if CHIP_CONFIG_SENDING_BATCH_COMMANDS_ENABLED
+    VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);
+
+    mRemoteMaxPathsPerInvoke = aRemoteMaxPathsPerInvoke;
+    mBatchCommandsEnabled    = true;
+    return CHIP_NO_ERROR;
+#else
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+#endif
+}
+
+CHIP_ERROR CommandSender::PrepareCommand(const CommandPathParams & aCommandPathParams, bool aStartDataStruct,
+                                         AdditionalInvokeRequestElements aOptionalArgs)
 {
     ReturnErrorOnFailure(AllocateBuffer());
 
     //
-    // We must not be in the middle of preparing a command, or having prepared or sent one.
+    // We must not be in the middle of preparing a command, or have already sent InvokeReponseMessage.
     //
-    VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);
+    bool canAddAnotherCommand = mState == State::AddedCommand && mBatchCommandsEnabled;
+    VerifyOrReturnError(mState == State::Idle || canAddAnotherCommand, CHIP_ERROR_INCORRECT_STATE);
+    // TODO what is the best error to give here? I thought CHIP_ERROR_SENTINEL might be an indicator for the called to then
+    // batch call send such that we can then try sending another InvokeRequest once we have processed all of Responses.
+    VerifyOrReturnError(mFinishedCommandCount < mRemoteMaxPathsPerInvoke, CHIP_ERROR_SENTINEL);
```

Comment:
Just want to clarify the use pattern here - is PrepareCommand called once for the entire batch, or once per command? The check here seems to imply that it's once per command. But lower down, aOptionalArgs is saved in the class member. If you call this multiple times, do you end up only tracking the last aOptionalArgs? or are these required to be the same or....

If you could add a comment about the expected use pattern in the header, it would probably help.

Reply:
You are calling this once per command. The order of operations is
```
PrepareCommand(...)
TLV::TLVWriter * writer = GetCommandDataIBTLVWriter()
// put things into the writer that is the contents of CommandDataIB
FinishCommand(...)
```

Most common use case if for something to just call `AddRequestData` which wraps everything for you, but there are cases of things calling the sequence described above themselves.

Arguably we could have this info provided in the finish, and save the ram. I will make that change.

My understanding, which might be incorrect, is that things outside of the SDK rely on the API of `CommandSender` to be backwards compatible. It would be nice to future proof this a little by passing in a parameter of a struct where the struct itself can change and you don't have parameters that start growing all with different defaults